discovery_engine_core/providers/src/gnews.rs

```diff
@@ -14,176 +14,105 @@
 
 //! Client to get new documents.
 
-use std::time::Duration;
-
+use async_trait::async_trait;
+use derive_more::From;
 use itertools::Itertools;
-use url::Url;
+use url::{form_urlencoded, Url, UrlQuery};
 
 use crate::{
-    filter::{Filter, Market},
+    rest::SimpleEndpoint,
     Article,
+    CommonQueryParts,
     Error,
+    HeadlinesProvider,
+    HeadlinesQuery,
+    NewsProvider,
+    NewsQuery,
 };
 
 use self::models::Response;
 
 mod models;
 
-const LATEST_HEADLINE_ENDPOINT: &str = "latest-headlines";
-const SEARCH_NEWS_ENDPOINT: &str = "search-news";
-
-/// Query parameters for the news search query
-pub struct NewsQuery<'a> {
-    /// Market of news.
-    pub market: Option<&'a Market>,
-    /// How many articles to return (per page).
-    pub page_size: usize,
-    /// The number of the page which should be returned.
-    ///
-    /// Paging starts with `1`.
-    pub page: usize,
-    /// Exclude given sources.
-    pub excluded_sources: &'a [String],
-    /// Search filter for selecting the results
-    pub filter: &'a Filter,
-}
-
-/// Query parameters for the headlines query
-pub struct HeadlinesQuery<'a> {
-    /// Market of news.
-    pub market: Option<&'a Market>,
-    /// How many articles to return (per page).
-    pub page_size: usize,
-    /// The number of the page which should be returned.
-    ///
-    /// Paging starts with `1`.
-    pub page: usize,
-    /// Exclude given sources.
-    pub excluded_sources: &'a [String],
-    /// Search filter for selecting the results
-    pub filter: Option<&'a Filter>,
-}
+/// Gnews based implementation of a `NewsProvider`.
+#[derive(From)]
+pub struct NewsProviderImpl(SimpleEndpoint);
```

Comment:
But it's not a client, it uses one. And if we would rename it it would still be `NewsClientImpl` because then we also would have to rename the trait (as they are named after the trait). And I would say `NewsClientImpl` is as unwieldy as `NewsProviderImpl`.

As a side note outside of `providers` all code except the multiplexer will use `dyn NewsProvider`.